packages/std/src/imports.rs

```diff
@@ -106,37 +106,39 @@ impl ExternalStorage {
 
 impl Storage for ExternalStorage {
     fn get(&self, key: &[u8]) -> Option<Vec<u8>> {
-        let key = build_region(key);
-        let key_ptr = &*key as *const Region as u32;
+        let key = Region::from_data(key);
+        let key_ptr = key.as_ptr() as u32;
 
         let read = unsafe { db_read(key_ptr) };
         if read == 0 {
             // key does not exist in external storage
             return None;
         }
 
-        let value_ptr = read as *mut Region;
-        let data = unsafe { consume_region(value_ptr) };
-        Some(data)
+        let value_ptr = read as *mut Region<Owned>;
+        let data = unsafe { Region::from_heap_ptr(value_ptr) };
+
+        Some(data.into_inner())
     }
 
     fn set(&mut self, key: &[u8], value: &[u8]) {
         if value.is_empty() {
             panic!("TL;DR: Value must not be empty in Storage::set but in most cases you can use Storage::remove instead. Long story: Getting empty values from storage is not well supported at the moment. Some of our internal interfaces cannot differentiate between a non-existent key and an empty value. Right now, you cannot rely on the behaviour of empty values. To protect you from trouble later on, we stop here. Sorry for the inconvenience! We highly welcome you to contribute to CosmWasm, making this more solid one way or the other.");
         }
 
-        // keep the boxes in scope, so we free it at the end (don't cast to pointers same line as build_region)
-        let key = build_region(key);
-        let key_ptr = &*key as *const Region as u32;
-        let mut value = build_region(value);
-        let value_ptr = &mut *value as *mut Region as u32;
+        let key = Region::from_data(key);
+        let key_ptr = key.as_ptr() as u32;
+
+        let value = Region::from_data(value);
+        let value_ptr = value.as_ptr() as u32;
+
         unsafe { db_write(key_ptr, value_ptr) };
     }
 
     fn remove(&mut self, key: &[u8]) {
-        // keep the boxes in scope, so we free it at the end (don't cast to pointers same line as build_region)
-        let key = build_region(key);
-        let key_ptr = &*key as *const Region as u32;
+        let key = Region::from_data(key);
```

Comment:
As mentioned below, here we could have a more specialized name that makes clear we don't copy that data into a vec or something live that.

Reply:
Yeah, we can decompose that into `from_bytes` and `from_vec`. Also gets rid of the annoying generics on the `from_data` function.